cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -174,5 +175,18 @@ Status ParquetFileFormat::ScanFile(const FileSource& source,
                                        out);
 }
 
+Status ParquetFileFormat::MakeFragment(const FileSource& source,
+                                       std::shared_ptr<ScanOptions> opts,
+                                       std::unique_ptr<FileBasedDataFragment>* out) {
+  fs::FileStats stats;
+  RETURN_NOT_OK(source.filesystem()->GetTargetStats(source.path(), &stats));
```

Comment:
Are you double-checking here? i.e. where is this method supposed to be call in the scanning pipeline?

Reply:
It is called to lazily generate a fragment from cached paths in a FileSystemBasedDataSource:
https://github.com/apache/arrow/pull/5139/files/7ad746cec7693aeea130d2f04ccb1232bb1ba176#diff-49c0202173daaea224a37a6d4e64f816R96

This call to GetTargetStats is used to check for file existence, the previous call to GetTargetStats is used to get a directory listing.

Reply:
Well, you'll still get a race condition if the file is removed just after the second `GetTargetStats`. So I don't think this check is necessary.

Reply:
I'll remove it